File: backtracking/permutations-ii/permutations-ii.ts
function permuteUnique(nums: number[]): number[][] {
  const ans: number[][] = [];
  const list: number[] = [];
  const visited: Set<number> = new Set();

  nums.sort((a, b) => a - b);

  backtrack(nums, visited, ans, list);

  return ans;
}

function backtrack(nums: number[], visited: Set<number>, ans: number[][], list: number[]) {
  if (list.length == nums.length) {
    ans.push(list.slice());
    return;
  }

  for (let i = 0; i < nums.length; i++) {
    if (visited.has(i)) {
      continue;
    }

    visited.add(i);
    list.push(nums[i]);

    backtrack(nums, visited, ans, list);

    list.pop();
    visited.delete(i);

    while (i < nums.length - 1 && nums[i] == nums[i + 1]) {
      i++;
    }
  }
}

File: backtracking/permutations-ii/permutations-ii.java
class Solution {
  public List<List<Integer>> permuteUnique(int[] nums) {
    List<List<Integer>> ans = new ArrayList<>();
    List<Integer> list = new ArrayList<>();
    Set<Integer> visited = new HashSet<>();
    Arrays.sort(nums);

    backtrack(nums, visited, ans, list);

    return ans;
  }

  public void backtrack(int[] nums, Set<Integer> visited, List<List<Integer>> ans, List<Integer> list) {

    if (list.size() == nums.length) {
      ans.add(new ArrayList<>(list));
      return;
    }

    for (int i = 0; i < nums.length; i++) {

      if (visited.contains(i)) {
        continue;
      }

      visited.add(i);
      list.add(nums[i]);
      backtrack(nums, visited, ans, list);

      list.remove(list.size() - 1);
      visited.remove(i);

      while (i < nums.length - 1 && nums[i] == nums[i + 1]) {
        i++;
      }
    }
  }
}
